[Promise] Add a C API for promises

tlively opened the pull request:
Add a new header, emscripten/promise.h, providing a C API for creating,
resolving, chaining, and combining promises from C with a focus on providing the
full expressive capability of JS promises. To provide maximum flexibility to
users, promises can be fulfilled or rejected with arbitrary user-provided
`void*` pointers as the value or reason, or they can be resolved by providing
the handle of another promise whose eventual state should be matched. The
callbacks provided to `emscripten_promise_then` additionally receive a context
`data` pointer provided by the user as part of the `emscripten_promise_then`
call.

Remove the less expressive C API for creating promises we had in eventloop.h.
That API allowed promises to be created and resolved, but was more complicated
and less capable.

tlively commented:
Current dependencies on/for this PR:
* main
  * **PR #18598** <a href="https://app.graphite.dev/github/pr/emscripten-core/emscripten/18598" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/emscripten-core/emscripten/18598?utm_source=stack-comment).

kripken commented on src/library.js:
```diff
@@ -588,9 +588,11 @@ mergeInto(LibraryManager.library, {
+      return f();
+    } finally {
+      stackRestore(stack);
+    }
```
Is this necessary with this PR? It might be worth doing as a separate change if so. (there might be performance implications here)

tlively replied:
Yes, because a user callback attached to a promise may throw, resulting in a rejected callback. We should be able to handle that gracefully and without corrupting the stack pointer. The alternative to this change would be doing the same thing in a one-off specialized utility in the library_promise.js code.

I suppose I should probably add tests for this as well.

tlively replied:
I no longer modify or use withStackSave in the latest version.

kripken commented on src/library_promise.js:
```diff
@@ -0,0 +1,146 @@
+        {{{ makeSetValue('resultPtr', 0, '0', '*') }}};
+        var result;
+        callUserCallback(() => {
+          result = {{{makeDynCall('ippp', 'callback')}}}(
```
```suggestion
          result = {{{ makeDynCall('ippp', 'callback') }}}(
```

kripken commented on src/library_promise.js:
```diff
@@ -0,0 +1,146 @@
+      dbg('emscripten promise callback: ' + value);
+#endif
+      return withStackSave(() => {
+        var resultPtr = stackAlloc(8);
```
Why 8?

tlively replied:
Ah, I should update this to `POINTER_SIZE`.

kripken commented on src/library_promise.js:
```diff
@@ -0,0 +1,146 @@
+          return getPromise(resultVal);
+        } else if (result == {{{ cDefine('EM_PROMISE_MATCH_RELEASE') }}}) {
+          var ret = getPromise(resultVal);
+          _emscripten_promise_destroy(resultVal);
```
this should be added to the `__deps`

kripken commented on system/include/emscripten/promise.h:
```diff
@@ -0,0 +1,104 @@
+// other function that returns a fresh promise, such as
+// `emscripten_promise_then`. It is fine to call `emscripten_promise_destroy` on
+// a promise handle before the promise is resolved; the configured callbacks
+// will still be called.
```
How does that work? Don't the callbacks do something with the the handle in some way (which would no longer be valid)?

tlively replied:
No, the callbacks don't generally need the handle because they're installed directly on the underlying promise object. That underlying promise is kept alive by the JS runtime even if it removed from the handle map.

kripken commented on system/include/emscripten/promise.h:
```diff
@@ -0,0 +1,104 @@
+// resolved to match the eventual state of `value` interpreted as a promise
+// handle. Finally, if `result` is `EM_PROMISE_REJECT`, then the promise is
+// rejected with the given value. Promises not created by
+// `emscripten_promise_create` should not be passed to this function.
```
What is a promise *not* created by that function?

tlively replied:
A promise created by `emscripten_promise_then`, `emscripten_promise_all`, or any future API that returns a promise. Analogously, you don't get to manually `resolve()` a promise you get back from `fetch()`, only promises you explicitly create with `new Promise(...)`.

kripken commented on system/include/emscripten/promise.h:
```diff
@@ -0,0 +1,104 @@
+} em_promise_t;
+
+typedef enum em_promise_result_t {
+  EM_PROMISE_FULFILL,
```
Why "fulfill" and not "resolve" (the standard JS term)?

tlively replied:
Fulfill and resolve mean different things ([docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#:~:text=You%20will%20also,well.%20For%20example%3A)) and I precisely mean "fulfill" here.

sbc100 commented on src/library_promise.js:
```diff
@@ -0,0 +1,162 @@
+#else
+        if (typeof e !== 'number') {
+          throw 0;
+        }
```
I'm confused by this.. what if a JS exceptions is thrown.. don't we want to re-throw it here?

tlively replied:
Any value thrown from here will become the "reason" value for the rejection of the new promise returned from `emscripten_promise_then`. If we are in wasm64 and the reason value is not a bigint, we'll get a TypeError when it is converted to a pointer to pass to the next rejection handler. To avoid that type error, we have to avoid throwing anything except a valid pointer here.

sbc100 commented on src/library_promise.js:
```diff
@@ -0,0 +1,162 @@
+        // the stack first.
+        stackRestore(stack);
+      }
+      if (result == {{{ cDefine('EM_PROMISE_FULFILL') }}}) {
```
Use a `switch` here?

sbc100 commented on src/library_promise.js:
```diff
@@ -0,0 +1,162 @@
+#if RUNTIME_DEBUG
+    err('emscripten_promise_resolve: ' + id);
+#endif
+    var info = promiseMap.get(id);
```
Use `switch` here?

sbc100 commented on src/library_promise.js:
```diff
@@ -0,0 +1,162 @@
+  },
+
+  emscripten_promise_all__deps: ['$promiseMap', '$getPromise', '$POINTER_SIZE'],
+  emscripten_promise_all__sig: 'pppp',
```
I think this returns just and integer, right?  So `ippp`?

tlively replied:
No, it returns a new promise handle which is a `void*` to drastically reduce the number of integer-to-pointer casts necessary to use this API.

sbc100 commented on src/library_promise.js:
```diff
@@ -0,0 +1,162 @@
+        if (resultBuf) {
+          for (var i = 0; i < size; i++) {
+            var result = results[i];
+            {{{ makeSetValue('resultBuf', 'i*POINTER_SIZE', 'result', '*') }}};
```
I think we prefer to use the compile time POINTER_SIZE.  Its pretty odd but in other places we do `i*${Runtime.POINTER_SIZE}` here

sbc100 replied:
(in backticks).

sbc100 commented on system/include/emscripten/promise.h:
```diff
@@ -0,0 +1,109 @@
+// An opaque handle to a JS Promise object.
+typedef struct em_promise_t {
+  void* handle;
+} em_promise_t;
```
How about just `typedef uint32_t em_promise_t`  ?   Or `typedef void* em_promise_t`  (i.e. Why does it need to be struct type?)

tlively replied:
I originally used `uint32_t` for the inner handle, but using the API often requires casting these handles to `void*`, so my test code was covered with ugly `*result = (void*)(uintptr_t)promise.handle;` lines. I changed the handle to be void* so that those lines can be much nicer: `*result = promise.handle`.

sbc100 replied:
Maybe put those casts into helpers/macros to if you think they ugly?  Are those casts needed when you want pass a promise as user data?   Aren't those cases expected whenever one wants to pass non-pointer-sized data as user data?   One could make the argument for make all types pointer sized based on that argument couldn't you?

I worry that we it will hard to change the underlying type here once we decide on its width.   I guess there is any way to avoid exposing its width.

sbc100 replied:
Also, why `*result = (void*)(uintptr_t)promise.handle` and hot `*result = (void*)(uintptr_t)promise`?  Isn't the em_promise type itself that thing you want to pass as the user data, and not the inner handle?

tlively replied:
Because I don't believe it's possible to cast a struct to a number or pointer, even if they're the same size (but I haven't checked).

tlively replied:
> Are those casts needed when you want pass a promise as user data?

Yes, but more importantly when you want to return from a promise callback with `EM_PROMISE_MATCH` or `EM_PROMISE_MATCH_RELEASE` and need to write the promise handle to the `result` pointer.

sbc100 replied:
Hmm.. makes me think maybe we use different types argument for different result types.  e.g and extra paramater called such as `em_handle_t* match_result_or_null`? 

Maybe its fine, but all that casting seems dirty.

tlively replied:
Macros or helpers could help, but I would prefer to avoid them as well. What would you think of doing `typedef struct _em_promise* em_promise_t` similar to how `pthread_t` works for C? For C at least, that would avoid the need for any casts and also avoid `.handle`, since all pointers are assignable to `void*`. C++ would still need explicit casts to `void*`, but that's probably ok (or we can do a different typedef in C++, like `pthread_t` has).

sbc100 commented on system/include/emscripten/promise.h:
```diff
@@ -0,0 +1,109 @@
+// promise will be rejected with that number converted to a pointer as its
+// rejection reason. If the callback throws any other value, the new promise
+// will be rejected with a NULL rejection reason.
+typedef em_promise_result_t (*promise_callback_t)(void** result,
```
`em_promise_callback_t` ?

sbc100 commented:
I don't mean to hold up this PR on cosmetic details.  I do suggest we perhaps mark it as experimental in some way so that we can continue to iterate on it as needed.

sbc100 commented:
(Perhaps just a comment at the top of the header)

sbc100 reviewed (approved):
lgtm, but I also just grabbed the `_destroy` part and integrated it into #18376, so I'm not actually blocked on this landed.

tlively commented:
Added comment about this being experimental and updated the typedef to be similar to `pthread_t`. I'll let this land as-is (if the tests pass), but we'll iterate as necessary.
